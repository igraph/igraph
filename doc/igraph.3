.\" Hey, Emacs!  This is an -*- nroff -*- source file.
.\"
.\" Copyright (C) 2006-2021  The igraph development team
.\"
.\" This is free software; you can redistribute it and/or modify it under
.\" the terms of the GNU General Public License as published by the Free
.\" Software Foundation; either version 2, or (at your option) any later
.\" version.
.\"
.\" This is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\" You should have received a copy of the GNU General Public License with
.\" your Debian GNU/Linux system, in /usr/share/common-licenses/GPL, or with
.\" the dpkg source package as the file COPYING.  If not, write to the Free
.\" Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH IGRAPH 3 "May 2021" "igraph library"
.SH NAME
igraph \- a library for creating and manipulating graphs
.SH DESCRIPTION
.B igraph
is a C library for complex network analysis and graph theory, with emphasis on
efficiency, portability and ease of use.
.SH DOCUMENTATION
The full documentation can be downloaded from the homepage of the
library:
.RI < https://igraph.org/c/doc >
.SH BUGS
If you think you have found a bug in igraph, feel free to file a bug report
in the issue tracker at:
.RI < https://github.com/igraph/igraph/issues >

.SH AUTHORS
Gabor Csardi <csardi.gabor@gmail.com>,
.br
Tamas Nepusz <ntamas@gmail.com>,
.br
Vincent Traag <vtraag@gmail.com>,
.br
Szabolcs Horvat <szhorvat@gmail.com>,
.br
Fabio Zanini <fabio.zanini@fastmail.fm>,
.br
Daniel Noom <ggatw@outlook.com>
