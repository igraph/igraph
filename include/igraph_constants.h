/* -*- mode: C -*-  */
/*
   IGraph library.
   Copyright (C) 2009-2012  Gabor Csardi <csardi.gabor@gmail.com>
   334 Harvard street, Cambridge, MA 02139 USA

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc.,  51 Franklin Street, Fifth Floor, Boston, MA
   02110-1301 USA

*/

#ifndef IGRAPH_CONSTANTS_H
#define IGRAPH_CONSTANTS_H

#include "igraph_config.h"
#include "igraph_decls.h"

__BEGIN_DECLS

/* -------------------------------------------------- */
/* Constants                                          */
/* -------------------------------------------------- */

typedef enum { IGRAPH_UNDIRECTED = 0, IGRAPH_DIRECTED = 1 } igraph_i_directed_t;

/* Note for the enum below: yes, IGRAPH_LOOPS_TWICE is 1, and IGRAPH_LOOPS_ONCE
 * is 2. This is intentional, for the sake of backwards compatibility with
 * earlier versions where we only had IGRAPH_LOOPS and it meant
 * IGRAPH_LOOPS_TWICE */
typedef enum { IGRAPH_NO_LOOPS = 0, IGRAPH_LOOPS = 1, IGRAPH_LOOPS_TWICE = 1, IGRAPH_LOOPS_ONCE = 2 } igraph_loops_t;

typedef enum { IGRAPH_NO_MULTIPLE = 0, IGRAPH_MULTIPLE = 1 } igraph_multiple_t;

typedef enum { IGRAPH_ASCENDING = 0, IGRAPH_DESCENDING = 1 } igraph_order_t;

typedef enum { IGRAPH_MINIMUM = 0, IGRAPH_MAXIMUM = 1 } igraph_optimal_t;

/* Do not renumber the following values! Some internal code treats them as bitmasks
 * and assumes that IGRAPH_ALL == IGRAPH_IN | IGRAPH_OUT and IGRAPH_IN & IGRAPH_OUT == 0. */
typedef enum { IGRAPH_OUT = 1, IGRAPH_IN = 2, IGRAPH_ALL = 3 } igraph_neimode_t;

/* Reverse IGRAPH_OUT to IGRAPH_IN and vice versa. Leave other values alone. */
#define IGRAPH_REVERSE_MODE(mode) \
    ((mode) == IGRAPH_IN ? IGRAPH_OUT : ((mode) == IGRAPH_OUT ? IGRAPH_IN : (mode)))

typedef enum { IGRAPH_WEAK = 1, IGRAPH_STRONG = 2 } igraph_connectedness_t;

typedef enum { IGRAPH_RECIPROCITY_DEFAULT = 0,
               IGRAPH_RECIPROCITY_RATIO = 1
             } igraph_reciprocity_t;

typedef enum { IGRAPH_ADJ_DIRECTED = 0,
               IGRAPH_ADJ_UNDIRECTED,
               IGRAPH_ADJ_UPPER, IGRAPH_ADJ_LOWER, IGRAPH_ADJ_MIN,
               IGRAPH_ADJ_PLUS,
               IGRAPH_ADJ_MAX,
             } igraph_adjacency_t;

typedef enum { IGRAPH_STAR_OUT = 0, IGRAPH_STAR_IN,
               IGRAPH_STAR_UNDIRECTED,
               IGRAPH_STAR_MUTUAL
             } igraph_star_mode_t;

typedef enum { IGRAPH_WHEEL_OUT = 0, IGRAPH_WHEEL_IN,
               IGRAPH_WHEEL_UNDIRECTED,
               IGRAPH_WHEEL_MUTUAL
             } igraph_wheel_mode_t;

typedef enum { IGRAPH_TREE_OUT = 0, IGRAPH_TREE_IN,
               IGRAPH_TREE_UNDIRECTED
             } igraph_tree_mode_t;

typedef enum { IGRAPH_ERDOS_RENYI_GNP = 0,
               IGRAPH_ERDOS_RENYI_GNM
             } igraph_erdos_renyi_t;

typedef enum { IGRAPH_GET_ADJACENCY_UPPER = 0,
               IGRAPH_GET_ADJACENCY_LOWER,
               IGRAPH_GET_ADJACENCY_BOTH
             } igraph_get_adjacency_t;

typedef enum { IGRAPH_DEGSEQ_CONFIGURATION = 0,     /* Configuration model, allowing non-simple graphs */
               IGRAPH_DEGSEQ_VL,                    /* Viger-Latapy, generates simple connected graphs */
               IGRAPH_DEGSEQ_FAST_HEUR_SIMPLE,      /* Fast heuristic, generates simple graphs */
               IGRAPH_DEGSEQ_CONFIGURATION_SIMPLE,  /* Configuration model, generates simple graphs */
               IGRAPH_DEGSEQ_EDGE_SWITCHING_SIMPLE, /* Edge-switching MCMC, generates simple graphs */

               /* Deprecated, kept for backwards compatibility: */
               IGRAPH_DEGSEQ_SIMPLE IGRAPH_DEPRECATED_ENUMVAL = IGRAPH_DEGSEQ_CONFIGURATION,
               IGRAPH_DEGSEQ_SIMPLE_NO_MULTIPLE IGRAPH_DEPRECATED_ENUMVAL = IGRAPH_DEGSEQ_FAST_HEUR_SIMPLE,
               IGRAPH_DEGSEQ_SIMPLE_NO_MULTIPLE_UNIFORM IGRAPH_DEPRECATED_ENUMVAL = IGRAPH_DEGSEQ_CONFIGURATION_SIMPLE
             } igraph_degseq_t;

typedef enum { IGRAPH_REALIZE_DEGSEQ_SMALLEST = 0,
               IGRAPH_REALIZE_DEGSEQ_LARGEST,
               IGRAPH_REALIZE_DEGSEQ_INDEX
             } igraph_realize_degseq_t;

typedef enum { IGRAPH_RANDOM_TREE_PRUFER = 0,
               IGRAPH_RANDOM_TREE_LERW
             } igraph_random_tree_t;

typedef enum { IGRAPH_FILEFORMAT_EDGELIST = 0,
               IGRAPH_FILEFORMAT_NCOL,
               IGRAPH_FILEFORMAT_PAJEK,
               IGRAPH_FILEFORMAT_LGL,
               IGRAPH_FILEFORMAT_GRAPHML
             } igraph_fileformat_type_t;

typedef enum { IGRAPH_REWIRING_SIMPLE = 0,
               IGRAPH_REWIRING_SIMPLE_LOOPS
             } igraph_rewiring_t;

typedef enum { IGRAPH_EDGEORDER_ID = 0,
               IGRAPH_EDGEORDER_FROM,
               IGRAPH_EDGEORDER_TO
             } igraph_edgeorder_type_t;

typedef enum { IGRAPH_TO_DIRECTED_ARBITRARY = 0,
               IGRAPH_TO_DIRECTED_MUTUAL,
               IGRAPH_TO_DIRECTED_RANDOM,
               IGRAPH_TO_DIRECTED_ACYCLIC
             } igraph_to_directed_t;

typedef enum { IGRAPH_TO_UNDIRECTED_EACH = 0,
               IGRAPH_TO_UNDIRECTED_COLLAPSE,
               IGRAPH_TO_UNDIRECTED_MUTUAL
             } igraph_to_undirected_t;

typedef enum { IGRAPH_VCONN_NEI_ERROR = 0,
               IGRAPH_VCONN_NEI_NUMBER_OF_NODES,
               IGRAPH_VCONN_NEI_IGNORE,
               IGRAPH_VCONN_NEI_NEGATIVE
             } igraph_vconn_nei_t;

typedef enum { IGRAPH_SPINCOMM_UPDATE_SIMPLE = 0,
               IGRAPH_SPINCOMM_UPDATE_CONFIG
             } igraph_spincomm_update_t;

typedef enum { IGRAPH_DONT_SIMPLIFY = 0,
               IGRAPH_SIMPLIFY
             } igraph_lazy_adlist_simplify_t;

typedef enum { IGRAPH_TRANSITIVITY_NAN = 0,
               IGRAPH_TRANSITIVITY_ZERO
             } igraph_transitivity_mode_t;

typedef enum { IGRAPH_SPINCOMM_IMP_ORIG = 0,
               IGRAPH_SPINCOMM_IMP_NEG
             } igraph_spinglass_implementation_t;

typedef enum { IGRAPH_COMMCMP_VI = 0,
               IGRAPH_COMMCMP_NMI,
               IGRAPH_COMMCMP_SPLIT_JOIN,
               IGRAPH_COMMCMP_RAND,
               IGRAPH_COMMCMP_ADJUSTED_RAND
             } igraph_community_comparison_t;

typedef enum { IGRAPH_ADD_WEIGHTS_NO = 0,
               IGRAPH_ADD_WEIGHTS_YES,
               IGRAPH_ADD_WEIGHTS_IF_PRESENT
             } igraph_add_weights_t;

typedef enum { IGRAPH_BARABASI_BAG = 0,
               IGRAPH_BARABASI_PSUMTREE,
               IGRAPH_BARABASI_PSUMTREE_MULTIPLE
             } igraph_barabasi_algorithm_t;

typedef enum { IGRAPH_FAS_EXACT_IP = 0,
               IGRAPH_FAS_APPROX_EADES
             } igraph_fas_algorithm_t;

typedef enum { IGRAPH_SUBGRAPH_AUTO = 0,
               IGRAPH_SUBGRAPH_COPY_AND_DELETE,
               IGRAPH_SUBGRAPH_CREATE_FROM_SCRATCH
             } igraph_subgraph_implementation_t;

typedef enum { IGRAPH_IMITATE_AUGMENTED = 0,
               IGRAPH_IMITATE_BLIND,
               IGRAPH_IMITATE_CONTRACTED
             } igraph_imitate_algorithm_t;

typedef enum { IGRAPH_LAYOUT_GRID = 0,
               IGRAPH_LAYOUT_NOGRID,
               IGRAPH_LAYOUT_AUTOGRID
             } igraph_layout_grid_t;

typedef enum { IGRAPH_RANDOM_WALK_STUCK_ERROR = 0,
               IGRAPH_RANDOM_WALK_STUCK_RETURN
             } igraph_random_walk_stuck_t;

typedef enum { IGRAPH_VORONOI_FIRST = 0,
               IGRAPH_VORONOI_LAST,
               IGRAPH_VORONOI_RANDOM
             } igraph_voronoi_tiebreaker_t;

typedef enum { IGRAPH_ROW_MAJOR = 0,
               IGRAPH_COLUMN_MAJOR = 1
             } igraph_matrix_storage_t;

__END_DECLS

#endif
