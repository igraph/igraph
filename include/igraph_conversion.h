/* -*- mode: C -*-  */
/*
   IGraph library.
   Copyright (C) 2009-2012  Gabor Csardi <csardi.gabor@gmail.com>
   334 Harvard street, Cambridge, MA 02139 USA

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc.,  51 Franklin Street, Fifth Floor, Boston, MA
   02110-1301 USA

*/

#ifndef IGRAPH_CONVERSION_H
#define IGRAPH_CONVERSION_H

#include "igraph_decls.h"
#include "igraph_constants.h"
#include "igraph_types.h"
#include "igraph_datatype.h"
#include "igraph_spmatrix.h"
#include "igraph_matrix.h"
#include "igraph_sparsemat.h"
#include "igraph_attributes.h"

__BEGIN_DECLS

/* -------------------------------------------------- */
/* Conversion                                         */
/* -------------------------------------------------- */

DECLDIR int igraph_get_adjacency(const igraph_t *graph, igraph_matrix_t *res,
                                 igraph_get_adjacency_t type, igraph_bool_t eids);
DECLDIR int igraph_get_adjacency_sparse(const igraph_t *graph, igraph_spmatrix_t *res,
                                        igraph_get_adjacency_t type);

DECLDIR int igraph_get_stochastic(const igraph_t *graph,
                                  igraph_matrix_t *matrix,
                                  igraph_bool_t column_wise);

DECLDIR int igraph_get_stochastic_sparsemat(const igraph_t *graph,
        igraph_sparsemat_t *sparsemat,
        igraph_bool_t column_wise);

DECLDIR int igraph_get_edgelist(const igraph_t *graph, igraph_vector_t *res, igraph_bool_t bycol);

DECLDIR int igraph_to_directed(igraph_t *graph,
                               igraph_to_directed_t flags);
DECLDIR int igraph_to_undirected(igraph_t *graph,
                                 igraph_to_undirected_t flags,
                                 const igraph_attribute_combination_t *edge_comb);
DECLDIR int igraph_to_prufer(const igraph_t *graph, igraph_vector_int_t *prufer);

__END_DECLS

#endif
