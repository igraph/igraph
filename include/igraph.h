/* -*- mode: C -*-  */
/*
   IGraph library.
   Copyright (C) 2003-2012  Gabor Csardi <csardi.gabor@gmail.com>
   334 Harvard street, Cambridge, MA 02139 USA

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc.,  51 Franklin Street, Fifth Floor, Boston, MA
   02110-1301 USA

*/

#ifndef IGRAPH_H
#define IGRAPH_H

#ifndef _GNU_SOURCE
    #define _GNU_SOURCE 1
#endif

#include "igraph_version.h"
#include "igraph_memory.h"
#include "igraph_error.h"
#include "igraph_random.h"
#include "igraph_progress.h"
#include "igraph_statusbar.h"

#include "igraph_types.h"
#include "igraph_complex.h"
#include "igraph_vector.h"
#include "igraph_matrix.h"
#include "igraph_array.h"
#include "igraph_dqueue.h"
#include "igraph_stack.h"
#include "igraph_heap.h"
#include "igraph_psumtree.h"
#include "igraph_strvector.h"
#include "igraph_vector_ptr.h"
#include "igraph_spmatrix.h"
#include "igraph_sparsemat.h"
#include "igraph_qsort.h"

#include "igraph_constants.h"
#include "igraph_datatype.h"
#include "igraph_iterators.h"
#include "igraph_interface.h"
#include "igraph_constructors.h"
#include "igraph_games.h"
#include "igraph_microscopic_update.h"
#include "igraph_centrality.h"
#include "igraph_paths.h"
#include "igraph_components.h"
#include "igraph_structural.h"
#include "igraph_transitivity.h"
#include "igraph_neighborhood.h"
#include "igraph_topology.h"
#include "igraph_bipartite.h"
#include "igraph_cliques.h"
#include "igraph_layout.h"
#include "igraph_visitor.h"
#include "igraph_community.h"
#include "igraph_conversion.h"
#include "igraph_foreign.h"
#include "igraph_motifs.h"
#include "igraph_operators.h"
#include "igraph_flow.h"
#include "igraph_nongraph.h"
#include "igraph_cocitation.h"
#include "igraph_adjlist.h"
#include "igraph_attributes.h"
#include "igraph_blas.h"
#include "igraph_lapack.h"
#include "igraph_arpack.h"
#include "igraph_mixing.h"
#include "igraph_separators.h"
#include "igraph_cohesive_blocks.h"
#include "igraph_eigen.h"
#include "igraph_hrg.h"
#include "igraph_threading.h"
#include "igraph_interrupt.h"
#include "igraph_scg.h"
#include "igraph_matching.h"
#include "igraph_embedding.h"
#include "igraph_scan.h"
#include "igraph_graphlets.h"
#include "igraph_epidemics.h"
#include "igraph_lsap.h"
#include "igraph_coloring.h"

#endif
