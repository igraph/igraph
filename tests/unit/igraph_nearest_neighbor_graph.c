#include <igraph.h>
#include <math.h>
#include <string.h>
#include "igraph_constants.h"
#include "igraph_conversion.h"
#include "igraph_error.h"
#include "igraph_interface.h"
#include "igraph_matrix.h"
#include "igraph_matrix_list.h"
#include "igraph_spatial.h"
#include "test_utilities.h"

igraph_error_t RKNN_neighbors(igraph_real_t* arr, igraph_integer_t num_points, igraph_integer_t dims, igraph_integer_t neighbors, igraph_real_t cutoff) {

    igraph_t graph;
    igraph_matrix_t points;
    igraph_matrix_t adj_mat;
    igraph_metric_t metric = IGRAPH_METRIC_L2;
    igraph_matrix_t dist_mat;

    IGRAPH_MATRIX_INIT_FINALLY(&dist_mat, num_points, num_points);

    IGRAPH_FINALLY(igraph_matrix_destroy, &points);
    igraph_matrix_init_array(&points, &arr[0], num_points, dims, IGRAPH_ROW_MAJOR);

    for (igraph_integer_t start = 0; start < num_points-1; start++ ) {
        for (igraph_integer_t end = start + 1; end < num_points; end++) {
            igraph_real_t distance = 0;
            switch (metric) {
                case IGRAPH_METRIC_L2:
                    for (igraph_integer_t i = 0; i < dims; i++) {
                        distance += pow(MATRIX(points, start, i) - MATRIX(points, end, i), 2);
                    }
            }
            MATRIX(dist_mat, start,end) = distance;
            MATRIX(dist_mat, end,start) = distance;
        }
    } 


    igraph_nearest_neighbor_graph(&graph, &points, metric, neighbors, cutoff);
    print_matrix(&points);
    print_graph_canon(&graph);
    igraph_matrix_init(&adj_mat, 0, 0);
    IGRAPH_FINALLY(igraph_matrix_destroy, &adj_mat);
    IGRAPH_CHECK(igraph_get_adjacency(&graph, &adj_mat, IGRAPH_GET_ADJACENCY_BOTH, NULL, IGRAPH_NO_LOOPS));
    IGRAPH_FINALLY(igraph_destroy, &graph);

    for (igraph_integer_t start = 0; start < num_points; start ++) {

            igraph_real_t min_missing=INFINITY;
            igraph_real_t max_present=0;
        for (igraph_integer_t end = 0; end < num_points; end ++) {
            if (start == end) continue;
            if (MATRIX(adj_mat, start, end) == 1) {
                if (max_present < MATRIX(dist_mat, start, end)) max_present = MATRIX(dist_mat, start, end);
            } else {
                if (min_missing > MATRIX(dist_mat, start, end)) min_missing = MATRIX(dist_mat, start, end);
            }
        }
        IGRAPH_ASSERT(min_missing > max_present);
        IGRAPH_ASSERT(max_present <= cutoff*cutoff);
    }

    igraph_destroy(&graph);
    igraph_matrix_destroy(&adj_mat);
    igraph_matrix_destroy(&points);
    igraph_matrix_destroy(&dist_mat);
    IGRAPH_FINALLY_CLEAN(4);


    return IGRAPH_SUCCESS;
}

int main(void) {


    igraph_real_t points2d[10] = {
        12  , 8,
        8  , 6,
        5  , 12,
        10  , 1,
        12  , 2
    };
    igraph_real_t points3d[15] = {
        1,6,4,
        6,2,3,
        3,6,6,
        3,2,2,
        2,3,3
    };
    printf("2d 2 neighbors, cutoff 5\n");
    RKNN_neighbors(&points2d[0], 5, 2, 2, 5);
    printf("2d 1 neighbors, cutoff INFINITY\n");
    RKNN_neighbors(&points2d[0], 5, 2, 1, INFINITY);
    printf("2d unlimited neighbors, cutoff INFINITY\n");
    RKNN_neighbors(&points2d[0], 5, 2, 0, INFINITY);
    printf("2d unlimited neighbors, cutoff 7\n");
    RKNN_neighbors(&points2d[0], 5, 2, 0, 7);

    printf("3d, 2 neighbors, cutoff 4\n");
    RKNN_neighbors(&points3d[0], 5, 3, 2, 4);
    printf("3d, unlimited neighbors, cutoff 4\n");
    RKNN_neighbors(&points3d[0], 5, 3, 0, 4);
    printf("3d, 2 neighbors, cutoff INFINITY\n");
    RKNN_neighbors(&points3d[0], 5, 3, 2, INFINITY);
    printf("3d, unlimited neighbors, cutoff INFINITY\n");
    RKNN_neighbors(&points3d[0], 5, 3, 0, INFINITY);

    VERIFY_FINALLY_STACK();
    return 0;
}
