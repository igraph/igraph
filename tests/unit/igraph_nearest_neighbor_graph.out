# L2 Metric test suite:

1d 2 neighbors, cutoff 3
[       12
         8
         5
        10
        12 ]
directed: true
vcount: 5
edges: {
0 3
0 4
1 3
3 0
3 1
4 0
4 3
}

1d 1 neighbors, cutoff INFINITY
[       12
         8
         5
        10
        12 ]
directed: true
vcount: 5
edges: {
0 4
1 3
2 1
3 0
4 0
}

1d unlimited neighbors, cutoff INFINITY
[       12
         8
         5
        10
        12 ]
directed: true
vcount: 5
edges: {
0 1
0 2
0 3
0 4
1 0
1 2
1 3
1 4
2 0
2 1
2 3
2 4
3 0
3 1
3 2
3 4
4 0
4 1
4 2
4 3
}

1d unlimited neighbors, cutoff 7
[       12
         8
         5
        10
        12 ]
directed: true
vcount: 5
edges: {
0 3
0 4
1 3
3 0
3 1
3 4
4 0
4 3
}

2d 2 neighbors, cutoff 5
[       12        8
         8        6
         5       12
        10        1
        12        2 ]
directed: true
vcount: 5
edges: {
0 1
1 0
3 4
4 3
}

2d 1 neighbors, cutoff INFINITY
[       12        8
         8        6
         5       12
        10        1
        12        2 ]
directed: true
vcount: 5
edges: {
0 1
1 0
2 1
3 4
4 3
}

2d unlimited neighbors, cutoff INFINITY
[       12        8
         8        6
         5       12
        10        1
        12        2 ]
directed: true
vcount: 5
edges: {
0 1
0 2
0 3
0 4
1 0
1 2
1 3
1 4
2 0
2 1
2 3
2 4
3 0
3 1
3 2
3 4
4 0
4 1
4 2
4 3
}

2d unlimited neighbors, cutoff 7
[       12        8
         8        6
         5       12
        10        1
        12        2 ]
directed: true
vcount: 5
edges: {
0 1
0 4
1 0
1 2
1 3
1 4
2 1
3 1
3 4
4 0
4 1
4 3
}

2d 2 neighbors, cutoff 1.2, degenerate case
[        0        0
         1        0
        -1        0
         0        1
         0       -1 ]
directed: true
vcount: 5
edges: {
0 1
0 2
1 0
2 0
3 0
4 0
}

3d, 2 neighbors, cutoff 4
[        1        6        4
         6        2        3
         3        6        6
         3        2        2
         2        3        3 ]
directed: true
vcount: 5
edges: {
0 2
0 4
1 3
2 0
3 1
3 4
4 0
4 3
}

3d, unlimited neighbors, cutoff 4
[        1        6        4
         6        2        3
         3        6        6
         3        2        2
         2        3        3 ]
directed: true
vcount: 5
edges: {
0 2
0 4
1 3
2 0
3 1
3 4
4 0
4 3
}

3d, 2 neighbors, cutoff INFINITY
[        1        6        4
         6        2        3
         3        6        6
         3        2        2
         2        3        3 ]
directed: true
vcount: 5
edges: {
0 2
0 4
1 3
1 4
2 0
2 4
3 1
3 4
4 0
4 3
}

3d, unlimited neighbors, cutoff INFINITY
[        1        6        4
         6        2        3
         3        6        6
         3        2        2
         2        3        3 ]
directed: true
vcount: 5
edges: {
0 1
0 2
0 3
0 4
1 0
1 2
1 3
1 4
2 0
2 1
2 3
2 4
3 0
3 1
3 2
3 4
4 0
4 1
4 2
4 3
}

4d 2 neighbors, cutoff 5
[        1        6        4        4
         6        2        3        3
         3        6        6        5
         3        2        2        3
         2        3        3        4 ]
directed: true
vcount: 5
edges: {
0 2
0 4
1 3
1 4
2 0
2 4
3 1
3 4
4 0
4 3
}

4d 1 neighbors, cutoff INFINITY
[        1        6        4        4
         6        2        3        3
         3        6        6        5
         3        2        2        3
         2        3        3        4 ]
directed: true
vcount: 5
edges: {
0 2
1 3
2 0
3 4
4 3
}

4d unlimited neighbors, cutoff INFINITY
[        1        6        4        4
         6        2        3        3
         3        6        6        5
         3        2        2        3
         2        3        3        4 ]
directed: true
vcount: 5
edges: {
0 1
0 2
0 3
0 4
1 0
1 2
1 3
1 4
2 0
2 1
2 3
2 4
3 0
3 1
3 2
3 4
4 0
4 1
4 2
4 3
}

4d unlimited neighbors, cutoff 4
[        1        6        4        4
         6        2        3        3
         3        6        6        5
         3        2        2        3
         2        3        3        4 ]
directed: true
vcount: 5
edges: {
0 2
0 4
1 3
2 0
3 1
3 4
4 0
4 3
}

3d, no points
[ 0-by-3 ]
directed: true
vcount: 0
edges: {
}

0d, no points
[ 0-by-0 ]
directed: true
vcount: 0
edges: {
}

3d, 1 point
[        1        6        4 ]
directed: true
vcount: 1
edges: {
}

2d, zero neighbors
[       12        8
         8        6
         5       12
        10        1
        12        2 ]
directed: true
vcount: 5
edges: {
}

2d, zero cutoff
[       12        8
         8        6
         5       12
        10        1
        12        2 ]
directed: true
vcount: 5
edges: {
}

0d, 1 point should error

Fibonacci spiral with 25 points, 1 neighbor unlimited cutoff.
[        0        0
  -0.737369 -0.67549
  0.123639   1.4088
   1.05385 -1.37456
  -1.96943 0.348364
   1.88669  1.20016
  -0.635898 -2.36551
  -1.21945  2.34797
    2.6568 -0.97026
  -2.77304 -1.14467
   1.34032  2.86418
  0.992612  -3.1646
  -2.99718  1.73693
   3.52145 0.774182
  -2.15194 -3.06091
  -0.49772  3.84087
    3.0586 -2.57779
  -4.11958 -0.170358
   3.00731  2.99267
  -0.201344 -4.35425
  -2.86534  3.43363
   4.54165 -0.611073
  -3.85017 -2.67885
    1.0526  4.67889
   2.43568 -4.25058 ]
directed: true
vcount: 25
edges: {
0 1
1 0
2 0
3 8
4 1
5 13
6 14
7 2
8 3
9 17
10 18
11 19
12 20
13 5
14 6
15 7
16 8
17 9
18 10
19 11
20 12
21 13
22 14
23 15
24 16
}
# L1 Metric test suite:

1d 2 neighbors, cutoff 3
[       12
         8
         5
        10
        12 ]
directed: true
vcount: 5
edges: {
0 3
0 4
1 3
3 0
3 1
4 0
4 3
}

1d 1 neighbors, cutoff INFINITY
[       12
         8
         5
        10
        12 ]
directed: true
vcount: 5
edges: {
0 4
1 3
2 1
3 0
4 0
}

1d unlimited neighbors, cutoff INFINITY
[       12
         8
         5
        10
        12 ]
directed: true
vcount: 5
edges: {
0 1
0 2
0 3
0 4
1 0
1 2
1 3
1 4
2 0
2 1
2 3
2 4
3 0
3 1
3 2
3 4
4 0
4 1
4 2
4 3
}

1d unlimited neighbors, cutoff 7
[       12
         8
         5
        10
        12 ]
directed: true
vcount: 5
edges: {
0 3
0 4
1 3
3 0
3 1
3 4
4 0
4 3
}

2d 2 neighbors, cutoff 5
[       12        8
         8        6
         5       12
        10        1
        12        2 ]
directed: true
vcount: 5
edges: {
3 4
4 3
}

2d 1 neighbors, cutoff INFINITY
[       12        8
         8        6
         5       12
        10        1
        12        2 ]
directed: true
vcount: 5
edges: {
0 1
1 0
2 1
3 4
4 3
}

2d unlimited neighbors, cutoff INFINITY
[       12        8
         8        6
         5       12
        10        1
        12        2 ]
directed: true
vcount: 5
edges: {
0 1
0 2
0 3
0 4
1 0
1 2
1 3
1 4
2 0
2 1
2 3
2 4
3 0
3 1
3 2
3 4
4 0
4 1
4 2
4 3
}

2d unlimited neighbors, cutoff 7
[       12        8
         8        6
         5       12
        10        1
        12        2 ]
directed: true
vcount: 5
edges: {
0 1
0 4
1 0
3 4
4 0
4 3
}

2d 2 neighbors, cutoff INFINITY, degenerate case
[        0        0
         1        0
        -1        0
         0        1
         0       -1 ]
directed: true
vcount: 5
edges: {
0 1
0 2
1 0
1 2
2 0
2 1
3 0
3 1
4 0
4 1
}

3d, 2 neighbors, cutoff 4
[        1        6        4
         6        2        3
         3        6        6
         3        2        2
         2        3        3 ]
directed: true
vcount: 5
edges: {
3 4
4 3
}

3d, unlimited neighbors, cutoff 4
[        1        6        4
         6        2        3
         3        6        6
         3        2        2
         2        3        3 ]
directed: true
vcount: 5
edges: {
3 4
4 3
}

3d, 2 neighbors, cutoff INFINITY
[        1        6        4
         6        2        3
         3        6        6
         3        2        2
         2        3        3 ]
directed: true
vcount: 5
edges: {
0 2
0 4
1 3
1 4
2 0
2 4
3 1
3 4
4 0
4 3
}

3d, unlimited neighbors, cutoff INFINITY
[        1        6        4
         6        2        3
         3        6        6
         3        2        2
         2        3        3 ]
directed: true
vcount: 5
edges: {
0 1
0 2
0 3
0 4
1 0
1 2
1 3
1 4
2 0
2 1
2 3
2 4
3 0
3 1
3 2
3 4
4 0
4 1
4 2
4 3
}

4d 2 neighbors, cutoff 5
[        1        6        4        4
         6        2        3        3
         3        6        6        5
         3        2        2        3
         2        3        3        4 ]
directed: true
vcount: 5
edges: {
1 3
3 1
3 4
4 3
}

4d 1 neighbors, cutoff INFINITY
[        1        6        4        4
         6        2        3        3
         3        6        6        5
         3        2        2        3
         2        3        3        4 ]
directed: true
vcount: 5
edges: {
0 2
1 3
2 0
3 1
4 3
}

4d unlimited neighbors, cutoff INFINITY
[        1        6        4        4
         6        2        3        3
         3        6        6        5
         3        2        2        3
         2        3        3        4 ]
directed: true
vcount: 5
edges: {
0 1
0 2
0 3
0 4
1 0
1 2
1 3
1 4
2 0
2 1
2 3
2 4
3 0
3 1
3 2
3 4
4 0
4 1
4 2
4 3
}

4d unlimited neighbors, cutoff 8
[        1        6        4        4
         6        2        3        3
         3        6        6        5
         3        2        2        3
         2        3        3        4 ]
directed: true
vcount: 5
edges: {
0 2
0 4
1 3
1 4
2 0
3 1
3 4
4 0
4 1
4 3
}
